### PR TITLE
Workaround to make PythonOps traced with torch.jit.trace work correctly.

### PR DESCRIPTION
**Pull request by zdevito**
The long-term fix is to remove the handling-creating pathways and
remove all the modes from PythonOp making it into an op that simply
calls a PyObject. Right now ONNX expects PythonOp to hold a
nn.Function, not a generic callable, so completely removing the legacy
pathway will also require changes to how ONNX symbolics are found.

### CODE REVIEW

**Comment by zdevito** [comment]
@jamesr66a @apaszke

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: This test depends on `Dropout` being implemented in Python. Implement your own autograd function instead, or take the test from my PR.

File: torch/csrc/jit/tracer_state.h
Comment (apaszke) [inline torch/csrc/jit/tracer_state.h]: Why would this ever be `true`? When do we need it?

File: torch/csrc/jit/tracer.h
Comment (apaszke) [inline torch/csrc/jit/tracer.h]: Just look how many places needed to be changed because the signature changed here. Since this is mostly a legacy thing, how about we remove it from there, and do a hacky `tracing_state->create_handles = true;` in those few places that need it.

File: torch/csrc/jit/tracer_state.h
Comment (zdevito) [reply torch/csrc/jit/tracer_state.h]: `torch.onnx.export` still uses the old tracing pathway where and we need do a bunch of yak shaving to remove it.

**Comment by zdevito** [comment]
Code is now updated. I had to change things around a bit because onnx export expects to be able to find ".symbolic" attributes, so we have to recover the original `Function` class in those cases. I added a `autogradFunction` object to do that.

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: Can you please use a custom Python function instead of Dropout? It's going to break those tests once someone moves it to C++ and it's not their fault.

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: That's not a legacy function. Can you please change the name?

File: torch/csrc/jit/tracer_state.h
Comment (apaszke) [reply torch/csrc/jit/tracer_state.h]: What do we need `creates_handles` for now? Looks like it's no longer necessary for ONNX, because it will correctly handle traces that had it set to `false`.

**Review by apaszke — COMMENTED** [review]
Can you remove `creates_handles` now? Looks like we don't have the ONNX dependency anymore

File: torch/csrc/jit/tracer_state.h
Comment (zdevito) [reply torch/csrc/jit/tracer_state.h]: We need it because I am pretty sure the ONNX symbolics expect those handles to be there, and I want to keep that fix, which will take a lot of iteration with the onnx tests, as a separate pr.

File: test/test_jit.py
Comment (zdevito) [reply test/test_jit.py]: There are two tests here. One tests if dropout export works, which was the original reason for this PR. The other is a more generic test that tracing a Python function from a new trace works. I updated the names but both are useful tests.

File: torch/csrc/jit/tracer_state.h
Comment (ezyang) [reply torch/csrc/jit/tracer_state.h]: I thought ONNX never supported handles at all, because it never supported backward export, and that's what the handles were originally added for. There is a bunch of special cases in the ONNX pass to make sure we handle handles correctly, but they're mostly checking if the handle is actually unused (which it should be if no backwards occurred.)

In any case if you guys need this to unblock I wouldn't block on this.

Comment (ezyang) [reply torch/csrc/jit/tracer_state.h]: (But it would be nice to have a comment saying this is temporary, and solely for supporting ONNX)

**Review by ezyang — APPROVED** [review]
But it would be nice to add references in PythonOp to the explanation of what it contains.

**Comment by ezyang** [comment]
Merge on green.

**Pull request merged**
